File: options.py
"""Options trading support for Public Terminal."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from decimal import Decimal


@dataclass
class OptionPosition:
    """Represents a single options contract position."""

    underlying_symbol: str
    option_type: str  # "CALL" or "PUT"
    strike_price: Decimal
    expiration_date: str  # ISO format: YYYY-MM-DD
    quantity: Decimal
    entry_price: Decimal
    current_value: Decimal | None = None
    last_price: Decimal | None = None
    contract_value: Decimal | None = None  # Usually 100 for stock options
    position_daily_gain: Decimal | None = None
    position_daily_gain_pct: Decimal | None = None

    @property
    def symbol_display(self) -> str:
        """Return human-readable contract symbol (e.g., 'AAPL 2026-05-16 150C')."""
        option_code = "C" if self.option_type.upper() == "CALL" else "P"
        return f"{self.underlying_symbol} {self.expiration_date} {self.strike_price}{option_code}"

    @property
    def days_to_expiry(self) -> int:
        """Calculate days remaining until expiration."""
        try:
            expiry = datetime.strptime(self.expiration_date, "%Y-%m-%d").date()
            today = datetime.now().date()
            return (expiry - today).days
        except (ValueError, TypeError):
            return 0

    @property
    def is_near_expiry(self) -> bool:
        """Check if option is within 7 days of expiration."""
        return 0 <= self.days_to_expiry <= 7

    def to_dict(self) -> dict:
        """Convert to dictionary for caching/display."""
        return {
            "underlying_symbol": self.underlying_symbol,
            "option_type": self.option_type,
            "strike_price": str(self.strike_price),
            "expiration_date": self.expiration_date,
            "quantity": str(self.quantity),
            "entry_price": str(self.entry_price),
            "current_value": str(self.current_value) if self.current_value else None,
            "last_price": str(self.last_price) if self.last_price else None,
            "contract_value": str(self.contract_value) if self.contract_value else None,
            "position_daily_gain": str(self.position_daily_gain) if self.position_daily_gain else None,
            "position_daily_gain_pct": str(self.position_daily_gain_pct) if self.position_daily_gain_pct else None,
        }

    @staticmethod
    def from_dict(data: dict) -> OptionPosition:
        """Reconstruct from dictionary."""
        return OptionPosition(
            underlying_symbol=data.get("underlying_symbol", ""),
            option_type=data.get("option_type", ""),
            strike_price=Decimal(str(data.get("strike_price", "0"))),
            expiration_date=data.get("expiration_date", ""),
            quantity=Decimal(str(data.get("quantity", "0"))),
            entry_price=Decimal(str(data.get("entry_price", "0"))),
            current_value=(
                Decimal(str(data.get("current_value")))
                if data.get("current_value")
                else None
            ),
            last_price=(
                Decimal(str(data.get("last_price")))
                if data.get("last_price")
                else None
            ),
            contract_value=(
                Decimal(str(data.get("contract_value")))
                if data.get("contract_value")
                else None
            ),
            position_daily_gain=(
                Decimal(str(data.get("position_daily_gain")))
                if data.get("position_daily_gain")
                else None
            ),
            position_daily_gain_pct=(
                Decimal(str(data.get("position_daily_gain_pct")))
                if data.get("position_daily_gain_pct")
                else None
            ),
        )


def _parse_occ_symbol(symbol: str) -> tuple[str, str, str, Decimal] | None:
    """
    Parse an OCC option symbol into (underlying, expiration_date, option_type, strike).

    OCC format: 6-char padded root + YYMMDD + C/P + 8-digit strike (thousandths).
    Example: "AAPL  260516C00150000" → ("AAPL", "2026-05-16", "CALL", Decimal("150"))
    Returns None if the symbol cannot be parsed.
    """
    s = symbol.replace(" ", "")
    # After stripping spaces the minimum valid length is 6(root)+6(date)+1(type)+8(strike)=21,
    # but root may be shorter once spaces are removed; find the C/P boundary from the right.
    # The last 15 chars are always YYMMDD + C/P + 8-digit strike.
    if len(s) < 15:
        return None
    date_type_strike = s[-15:]
    underlying = s[:-15].strip()
    try:
        yy, mm, dd = date_type_strike[:2], date_type_strike[2:4], date_type_strike[4:6]
        expiration_date = f"20{yy}-{mm}-{dd}"
        option_type = "CALL" if date_type_strike[6].upper() == "C" else "PUT"
        strike = Decimal(date_type_strike[7:]) / Decimal("1000")
    except (ValueError, IndexError):
        return None
    if not underlying:
        return None
    return underlying, expiration_date, option_type, strike


def extract_options_from_positions(positions) -> list[OptionPosition]:
    """
    Parse options positions from Public API portfolio.positions list.

    The SDK's PortfolioInstrument has only symbol/name/type — no details field.
    Option details are decoded from the OCC-format symbol string.
    """
    options = []

    for pos in positions or []:
        instrument = getattr(pos, "instrument", None)
        if not instrument or instrument.type.value != "OPTION":
            continue

        parsed = _parse_occ_symbol(instrument.symbol)
        if not parsed:
            continue
        underlying_symbol, expiration_date, option_type, strike_price = parsed

        quantity = Decimal(str(pos.quantity)) if pos.quantity else Decimal("0")
        current_value = pos.current_value or Decimal("0")

        last_price = None
        if pos.last_price and pos.last_price.last_price is not None:
            try:
                last_price = Decimal(str(pos.last_price.last_price))
            except (ValueError, TypeError):
                pass

        entry_price = Decimal("0")
        if pos.cost_basis and pos.cost_basis.unit_cost is not None:
            try:
                entry_price = Decimal(str(pos.cost_basis.unit_cost))
            except (ValueError, TypeError):
                pass

        daily_gain = None
        daily_gain_pct = None
        if pos.position_daily_gain:
            gain_obj = pos.position_daily_gain
            try:
                if gain_obj.gain_value is not None:
                    daily_gain = Decimal(str(gain_obj.gain_value))
                if gain_obj.gain_percentage is not None:
                    daily_gain_pct = Decimal(str(gain_obj.gain_percentage))
            except (ValueError, TypeError):
                pass

        options.append(
            OptionPosition(
                underlying_symbol=underlying_symbol,
                option_type=option_type,
                strike_price=strike_price,
                expiration_date=expiration_date,
                quantity=quantity,
                entry_price=entry_price,
                current_value=current_value,
                last_price=last_price,
                contract_value=Decimal("100"),
                position_daily_gain=daily_gain,
                position_daily_gain_pct=daily_gain_pct,
            )
        )

    return options

File: internal/options/options.go
// Package options parses OCC option symbols and extracts option positions from
// a portfolio. Direct port of options.py.
package options

import (
	"fmt"
	"strconv"
	"strings"
	"time"
	"unicode"

	"github.com/shopspring/decimal"

	"github.com/ks1686/public-terminal/internal/api"
)

// OptionPosition holds data for a single options contract position.
type OptionPosition struct {
	UnderlyingSymbol string
	OptionType       string // "CALL" or "PUT"
	StrikePrice      decimal.Decimal
	ExpirationDate   string // "YYYY-MM-DD"
	Quantity         decimal.Decimal
	CurrentValue     decimal.Decimal
	LastPrice        decimal.Decimal
	DailyGainPct     *decimal.Decimal
	DaysToExpiry     int
	OCCSymbol        string
}

// SymbolDisplay returns a compact representation like "AAPL 260516C150.00".
func (o OptionPosition) SymbolDisplay() string {
	parts := strings.Fields(o.OCCSymbol)
	if len(parts) == 0 {
		return o.OCCSymbol
	}
	underlying := parts[0]
	rest := ""
	if len(parts) > 1 {
		rest = strings.Join(parts[1:], "")
	}
	// Try to produce a friendly label from parsed fields
	if o.OptionType != "" && !o.StrikePrice.IsZero() {
		t := strings.ToUpper(string(o.OptionType[0]))
		exp := strings.ReplaceAll(o.ExpirationDate, "-", "")[2:] // YYMMDD
		return fmt.Sprintf("%s %s%s%.2f", underlying, exp, t, must(o.StrikePrice.Float64()))
	}
	return fmt.Sprintf("%s %s", underlying, rest)
}

func must(f float64, _ bool) float64 { return f }

// IsNearExpiry returns true when DaysToExpiry is 0–7 (inclusive).
func (o OptionPosition) IsNearExpiry() bool {
	return o.DaysToExpiry >= 0 && o.DaysToExpiry <= 7
}

// ToDict returns a map suitable for the TUI's options table cache.
func (o OptionPosition) ToDict() map[string]any {
	var gainStr string
	gainPositive := false
	if o.DailyGainPct != nil {
		pct, _ := o.DailyGainPct.Float64()
		sign := ""
		if pct >= 0 {
			sign = "+"
			gainPositive = true
		}
		gainStr = fmt.Sprintf("%s%.2f%%", sign, pct)
	} else {
		gainStr = "—"
	}
	var lastPriceStr, valueStr string
	if !o.LastPrice.IsZero() {
		lastPriceStr = fmt.Sprintf("$%.2f", must(o.LastPrice.Float64()))
	} else {
		lastPriceStr = "—"
	}
	if !o.CurrentValue.IsZero() {
		valueStr = fmt.Sprintf("$%.2f", must(o.CurrentValue.Float64()))
	} else {
		valueStr = "—"
	}
	valueNum, _ := o.CurrentValue.Float64()
	return map[string]any{
		"symbol_display":  o.SymbolDisplay(),
		"underlying":      o.UnderlyingSymbol,
		"type":            o.OptionType,
		"strike":          o.StrikePrice.StringFixed(2),
		"expiry":          o.ExpirationDate,
		"qty":             o.Quantity.String(),
		"price":           lastPriceStr,
		"value":           valueStr,
		"value_num":       valueNum,
		"gain":            gainStr,
		"gain_positive":   gainPositive,
		"days_to_expiry":  o.DaysToExpiry,
	}
}

// ─────────────────────────────────────────────────────────────────────────────
// OCC symbol parsing
// ─────────────────────────────────────────────────────────────────────────────

// parseOCCSymbol parses an OCC option symbol into its components.
// Format: "{Underlying}{YYMMDD}{C|P}{StrikePrice*1000:08d}"
// e.g. "AAPL  260516C00150000" → underlying="AAPL", expiry="2026-05-16", type="CALL", strike=150.00
func parseOCCSymbol(occ string) (underlying, optType, expiry string, strike decimal.Decimal, err error) {
	// Strip spaces and extract parts
	s := strings.TrimSpace(occ)
	// Find the transition from alpha (underlying) to digit (date)
	idx := -1
	for i, ch := range s {
		if unicode.IsDigit(ch) {
			idx = i
			break
		}
	}
	if idx < 1 || idx+6 > len(s) {
		return "", "", "", decimal.Zero, fmt.Errorf("invalid OCC symbol: %q", occ)
	}
	underlying = strings.TrimSpace(s[:idx])
	rest := s[idx:]
	if len(rest) < 15 {
		return "", "", "", decimal.Zero, fmt.Errorf("OCC symbol too short: %q", occ)
	}
	dateStr := rest[:6] // YYMMDD
	typeChar := rest[6]
	strikeStr := rest[7:15]

	year, _ := strconv.Atoi(dateStr[:2])
	month, _ := strconv.Atoi(dateStr[2:4])
	day, _ := strconv.Atoi(dateStr[4:6])
	fullYear := 2000 + year
	expiry = fmt.Sprintf("%04d-%02d-%02d", fullYear, month, day)

	switch typeChar {
	case 'C', 'c':
		optType = "CALL"
	case 'P', 'p':
		optType = "PUT"
	default:
		return "", "", "", decimal.Zero, fmt.Errorf("unknown option type %q in %q", string(typeChar), occ)
	}

	strikeInt, err := strconv.ParseInt(strikeStr, 10, 64)
	if err != nil {
		return "", "", "", decimal.Zero, fmt.Errorf("invalid strike in %q: %w", occ, err)
	}
	strike = decimal.NewFromInt(strikeInt).Div(decimal.NewFromInt(1000))
	return underlying, optType, expiry, strike, nil
}

func daysToExpiry(expiryDate string) int {
	t, err := time.Parse("2006-01-02", expiryDate)
	if err != nil {
		return -1
	}
	days := int(time.Until(t).Hours() / 24)
	return days
}

// ─────────────────────────────────────────────────────────────────────────────
// Extraction
// ─────────────────────────────────────────────────────────────────────────────

// ExtractOptionsFromPositions filters OPTION positions from a portfolio and
// returns parsed OptionPosition values.
func ExtractOptionsFromPositions(positions []api.Position) []OptionPosition {
	var out []OptionPosition
	for _, pos := range positions {
		if pos.Instrument.Type != "OPTION" {
			continue
		}
		occ := pos.Instrument.Symbol
		underlying, optType, expiry, strike, err := parseOCCSymbol(occ)
		if err != nil {
			// Fallback: surface the raw symbol
			underlying = occ
		}

		var lastPrice decimal.Decimal
		if pos.LastPrice != nil && pos.LastPrice.LastPrice != nil {
			lastPrice = *pos.LastPrice.LastPrice
		}

		var currentValue decimal.Decimal
		if pos.CurrentValue != nil {
			currentValue = *pos.CurrentValue
		}

		var gainPct *decimal.Decimal
		if pos.PositionDailyGain != nil && pos.PositionDailyGain.GainPercentage != nil {
			gainPct = pos.PositionDailyGain.GainPercentage
		}

		out = append(out, OptionPosition{
			UnderlyingSymbol: underlying,
			OptionType:       optType,
			StrikePrice:      strike,
			ExpirationDate:   expiry,
			Quantity:         pos.Quantity,
			CurrentValue:     currentValue,
			LastPrice:        lastPrice,
			DailyGainPct:     gainPct,
			DaysToExpiry:     daysToExpiry(expiry),
			OCCSymbol:        occ,
		})
	}
	return out
}
